Add parallel tests suggestion

thomas4019 opened the pull request:
Running tests in parallel is important for a large Node.js project.

i0natan commented:
@thomas4019 Welcome! Planning to review soon, looks very interesting

BrunoScheufler commented on README.md:
```diff
@@ -475,6 +475,12 @@ All statements above will return false if used with `===`
 
+## ![✔] 4.10 Run parallelized tests
+
+**TL;DR:** Consider running your tests with [Jest](https://jestjs.io), [Ava](https://github.com/avajs/ava) or another framework that supports parallelization out of the box. This will require you to write test files that don't conflict, but means your tests will execute substantially faster. You should also avoid tests with real IO (such as network or db calls), since they'll slow them down and m ...
```
```suggestion
**TL;DR:** Consider running your tests with [Jest](https://jestjs.io), [ava](https://github.com/avajs/ava) or another framework that supports parallelization out of the box. This approach requires you to write tests that don't conflict, resulting in your tests executing substantially faster. You should also avoid tests with real I/O (such as network or database calls), since they'll impact test performance and make parallelization harder. Note that it can be difficult to migrate existing tests written for Mocha to run in parallel, so if you expect your project to grow large, starting with parallel tests helps.
```

BrunoScheufler replied:
Great work 👍  I've added some small changes, you don't have to include them one-to-one, just some quick thoughts 😄

thomas4019 commented:
Hi @BrunoScheufler and @i0natan, I've incorporated your suggestions. I removed the note about I/O since I do think that's a separate issue. I also renamed the title and added more thoughts about keeping tests independent. Curious to know your thoughts, thanks!

i0natan commented:
@thomas4019 

We're progressing toward a great addition.

One important remark: parallelization and independent tests are completely different topics hence should be separated bullets. I suggest at first extracting out the 'independent tests' from this bullet (focus on parallelization) and separating it to a new bullet using another PR (might for now just extract out and write the new one later)

Also, the read more section links to some other bullet which deal with another topic.

I suggest editing before final review, cool?

See #276

thomas4019 commented:
@i0natan Making a different section for independent tests is a good idea. I've done that, improved the text a bit and put the more info link back into the right place.

i0natan commented:
@thomas4019 Thank you, shall review in 2 days. We might have 1-2 more iterations to perfect this important piece of advice.

p.s. I didn't know that AVA knows to parallel cases (not suits/files), interesting. I suggest you read a bit about CI parallelization - most vendors can run tests across multiple containers and machines

@BrunoScheufler Any chance you can help here with creating/showing how to put a 'read more' section?

i0natan commented:
/remind me in 2 days

i0natan commented on README.md:
```diff
@@ -477,6 +477,12 @@ All statements above will return false if used with `===`
 
 🔗 [**Read More: Refactoring!**](/sections/testingandquality/refactoring.md)
 
+## ![✔] 4.10 Run tests in parallel
```
Add the WHY - "Run tests in parallel for faster feedback loop"

(feedback loop is a term from the testing world)

#just-an-idea

i0natan commented on README.md:
```diff
@@ -477,6 +477,12 @@ All statements above will return false if used with `===`
+**TL;DR:** Design your tests so you can run them concurrently. Your total test times will be dramatically lower since the runner can take advantage of multiple CPU cores and slow tests won't block execution. Consider a framework that supports parallelization out of the box. Some run multiple tests in the same file in parallel ([Ava](https://github.com/avajs/ava)), whereas others only run multiple ...
+
+**Otherwise:** As your project grows larger your test execution times will continue getting longer and become unmanageable.
+
```
Each bullet must have a readmore file with content, [see number 4 here](https://github.com/i0natan/nodebestpractices/blob/master/.operations/writing-guidelines.md).

If need help, I can gladly add these file with template for you + ideas what to write about inside.

Basically, within the read-more file I would put the list of tools you mentioned, explain what CIs can do, code example antipattern (designing a test which can't be parallelized), valid code example, graph showing performance of various libraries, etc

#mandatory

i0natan commented on README.md:
```diff
@@ -477,6 +477,12 @@ All statements above will return false if used with `===`
 
+## ![✔] 4.10 Run tests in parallel
+
+**TL;DR:** Design your tests so you can run them concurrently. Your total test times will be dramatically lower since the runner can take advantage of multiple CPU cores and slow tests won't block execution. Consider a framework that supports parallelization out of the box. Some run multiple tests in the same file in parallel ([Ava](https://github.com/avajs/ava)), whereas others only run multiple ...
```
Why is it difficult to parallelize existing tests? this part deserves some clarification

#just-an-idea

i0natan commented on README.md:
```diff
@@ -477,6 +477,12 @@ All statements above will return false if used with `===`
 🔗 [**Read More: Refactoring!**](/sections/testingandquality/refactoring.md)
 
+## ![✔] 4.10 Run tests in parallel
+
```
Consider starting the section with the "so what" - why is it important, there are many sources on the testing feedback loop and why providing the developer with almost-immediate feedback is valuable

i0natan commented:
@thomas4019 Let's run 1-2 last improvement iterations and merge?

You may grab some more ideas for parallelization from my post here: https://medium.com/@me_37286/yoni-goldberg-javascript-nodejs-testing-best-practices-2b98924c9347
